### ISSUE TITLE
Pass query params/embeds to useGetOne

### ISSUE THREAD

**Issue by RWOverdijk** [marmelab/react-admin#7464]
**Is your feature request related to a problem? Please describe.**

I'm currently not able to pass in `_embed` for any of the components (<List />, <Edit />) or even useGetOne (json-server because I am using ra-data-json-server).

**Describe the solution you'd like**

At the core of this FR is that I'd like to be able to specify which fields to populate when using `<Edit />`. I can't find solutions for n-m relations. Yes, there's for example an `ArrayInput` but it doesn't change the behaviour of the getOne.

I find the way this seemingly works strange and hard to understand. Why would a `findOne` endpoint return an array of ids for relations? Why perform a join if it wasn't requested? Why would a findOne never want to add query params for embeds? Am I just doing something very wrong is or this an actual oversight?

**Practical example**

I do prefer talking in terms of practical examples so here's one:

I have roles and permissions. A role has a name and has multiple permissions. A permission has a resource, action and can belong to multiple roles.

`GET /roles` returns roles, but not the permissions. `GET /roles?_embed=permissions` does. Without the _embed, how can I populate the `<ReferenceArrayInput />`?

**Describe alternatives you've considered**

I looked into creating my own <Edit /> but following the trail to the adapter it still involves quite a bit of work.

Even when `meta` becomes available in components this means there's a need for a custom data provider. I think adding a `query` to getOne would offer the most flexibility.

**Additional context**

I am on the RC for react-admin 4.

I'm almost certain I am just missing something. But if I'm not, I think this would make for a useful feature.

**Update**

To verify if I understood I changed my api to return an array of ids and that does indeed allow me to work with the <Edit /> page. I've never seen an API perform joins like this before so I wonder if this is really what react-admin assumes or if I'm still missing something

**Comment by fzaninotto** [comment]
This is a fundamental aspect of the react-admin architecture, and it deserves an explanation. 

We've built it primarily for REST APIs, so it works *without* embeds and does the "joins" client-side. You don't need to *embed* the permissions in your roles to use `<ReferenceArrayInput>`, because `<ReferenceArrayInput>` will fetch the permissions based on an array of ids on mount. This call is optimized (i.e. all the `<ReferenceArrayInput>` of a given page are aggregated and deduplicated, so even in a list or roles this only triggers one call to `getMany('permissions')`).

This design choice allows developers to build user interfaces faster, without worrying about adding embed support server-side, or thinking in advance about the aggregated data they need for one page. It allows them to work with legacy APIs, or automated APIs based on a data model. It has a small performance cost, albeit painless in 90% of the cases as the `<ReferenceArrayInput>` doesn't block rendering. 

So what you ask for isn't necessary for react-admin to work.

Now, if you *want* to embed a relationship, either you do it for all views, in which case it's your responsibility to add the `_embed` parameter in the Data Provider URL for this resource, or you want to do it only for certain pages, in which case the `meta` parameter should do the trick. 

I hope that helps remove the confusion.

**Comment by RWOverdijk** [comment]
@fzaninotto Thanks for the explanation, but I'm still not sure I get it. Like I commented on the other issue (sorry for the spam 😅) this confusion is specifically for <Edit>

My interpretation of your explanation assumes that my API returns an array of ids for its relationships when no _embed was provided. But it doesn't do that to prevent performing unnecessary joins. So in order to get the ids, the server needs to know that the ids are needed. We don't use mongodb, so we don't have an array of ids. We have join tables that require another join as part of the query.

If including the IDs in the response is a requirement I'll adjust my API accordingly, I just want to make sure that this is the way react-admin assumes my API functions since I can't find that explanation explicitely anywhere.

TL;DR; Does <Edit> require me to return an array of IDs for relations in order to populate the current relations?
